# ML-07 — Baseline Action Score and Top-20 Review

[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/ihaseeb0081/flyrank-ml-internship-haseeb/blob/main/work/notebooks/w04_baseline_score.ipynb)

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

In [33]:
import os

print(os.listdir("/content"))

['.config', 'flyrank-ml-internship-haseeb', 'sample_data']


In [34]:
import os

for item in os.listdir("/content"):
    print(item)

.config
flyrank-ml-internship-haseeb
sample_data


In [35]:
import os

for root, dirs, files in os.walk("/content"):
    for file in files:
        if file.endswith(".csv"):
            print(os.path.join(root, file))

/content/flyrank-ml-internship-haseeb/outputs/refresh_queue_sample.csv
/content/flyrank-ml-internship-haseeb/data/raw/content_refresh_anonymized.csv
/content/flyrank-ml-internship-haseeb/work/outputs/baseline_action_score.csv
/content/sample_data/mnist_test.csv
/content/sample_data/mnist_train_small.csv
/content/sample_data/california_housing_train.csv
/content/sample_data/california_housing_test.csv


In [36]:
import pandas as pd
import os

print(os.listdir("/content"))

['.config', 'flyrank-ml-internship-haseeb', 'sample_data']


In [37]:
import os

for root, dirs, files in os.walk("/content"):
    for file in files:
        if file.endswith(".csv"):
            print(os.path.join(root, file))

/content/flyrank-ml-internship-haseeb/outputs/refresh_queue_sample.csv
/content/flyrank-ml-internship-haseeb/data/raw/content_refresh_anonymized.csv
/content/flyrank-ml-internship-haseeb/work/outputs/baseline_action_score.csv
/content/sample_data/mnist_test.csv
/content/sample_data/mnist_train_small.csv
/content/sample_data/california_housing_train.csv
/content/sample_data/california_housing_test.csv


In [38]:
import os

print("Files/folders in /content:")
print(os.listdir("/content"))

Files/folders in /content:
['.config', 'flyrank-ml-internship-haseeb', 'sample_data']


In [39]:
import os

print("CSV files found:")

for root, dirs, files in os.walk("/content"):
    for file in files:
        if file.lower().endswith(".csv"):
            print(os.path.join(root, file))

CSV files found:
/content/flyrank-ml-internship-haseeb/outputs/refresh_queue_sample.csv
/content/flyrank-ml-internship-haseeb/data/raw/content_refresh_anonymized.csv
/content/flyrank-ml-internship-haseeb/work/outputs/baseline_action_score.csv
/content/sample_data/mnist_test.csv
/content/sample_data/mnist_train_small.csv
/content/sample_data/california_housing_train.csv
/content/sample_data/california_housing_test.csv


In [40]:
import os

print(os.listdir("/content"))

['.config', 'flyrank-ml-internship-haseeb', 'sample_data']


In [41]:
!git clone "https://github.com/ihaseeb0081/flyrank-ml-internship-haseeb"

fatal: destination path 'flyrank-ml-internship-haseeb' already exists and is not an empty directory.


In [42]:
import os

print(os.listdir("/content"))

['.config', 'flyrank-ml-internship-haseeb', 'sample_data']


In [43]:
import os

for root, dirs, files in os.walk("/content"):
    for file in files:
        if file.lower().endswith(".csv"):
            print(os.path.join(root, file))

/content/flyrank-ml-internship-haseeb/outputs/refresh_queue_sample.csv
/content/flyrank-ml-internship-haseeb/data/raw/content_refresh_anonymized.csv
/content/flyrank-ml-internship-haseeb/work/outputs/baseline_action_score.csv
/content/sample_data/mnist_test.csv
/content/sample_data/mnist_train_small.csv
/content/sample_data/california_housing_train.csv
/content/sample_data/california_housing_test.csv


In [44]:
import os

print("CSV files in the repository:")

for root, dirs, files in os.walk("/content"):
    for file in files:
        if file.lower().endswith(".csv"):
            print(os.path.join(root, file))

CSV files in the repository:
/content/flyrank-ml-internship-haseeb/outputs/refresh_queue_sample.csv
/content/flyrank-ml-internship-haseeb/data/raw/content_refresh_anonymized.csv
/content/flyrank-ml-internship-haseeb/work/outputs/baseline_action_score.csv
/content/sample_data/mnist_test.csv
/content/sample_data/mnist_train_small.csv
/content/sample_data/california_housing_train.csv
/content/sample_data/california_housing_test.csv


In [45]:
import pandas as pd

df = pd.read_csv(
    "/content/flyrank-ml-internship-haseeb/data/raw/content_refresh_anonymized.csv"
)

print("Dataset shape:", df.shape)
display(df.head())

Dataset shape: (30000, 44)


,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


In [46]:
print("Columns in the dataset:")

for col in df.columns:
    print(col)

Columns in the dataset:
content_id
client_id
search_volume
competition
competition_level
cpc
content_type
main_intent
word_count
char_count
provider_used
model_used
impressions_90d
clicks_90d
pageviews_90d
sessions_90d
users_90d
engaged_sessions_90d
ai_sessions_90d
scroll_events_90d
days_with_impressions
days_with_sessions
impressions_last_30d
clicks_last_30d
sessions_last_30d
impressions_prev_30d
clicks_prev_30d
sessions_prev_30d
content_age_days
age_tier
age_tier_order
days_since_last_update
freshness_tier
word_count_tier
char_count_tier
ctr
avg_position
engagement_rate
scroll_rate
ai_traffic_pct
impression_tier
position_tier
trend_direction
trend_pct


In [47]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 44 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   content_id              30000 non-null  object 
 1   client_id               30000 non-null  object 
 2   search_volume           27532 non-null  float64
 3   competition             27532 non-null  float64
 4   competition_level       27390 non-null  object 
 5   cpc                     27532 non-null  float64
 6   content_type            30000 non-null  object 
 7   main_intent             27626 non-null  object 
 8   word_count              22301 non-null  float64
 9   char_count              22301 non-null  float64
 10  provider_used           8562 non-null   object 
 11  model_used              24267 non-null  object 
 12  impressions_90d         30000 non-null  int64  
 13  clicks_90d              30000 non-null  int64  
 14  pageviews_90d           30000 non-null

In [48]:
display(df.isna().sum().sort_values(ascending=False))

,0
provider_used,21438
word_count,7699
char_count,7699
word_count_tier,7699
char_count_tier,7699
model_used,5733
trend_pct,3388
competition_level,2610
search_volume,2468
cpc,2468


In [49]:
for col in df.columns:
    print(col)

content_id
client_id
search_volume
competition
competition_level
cpc
content_type
main_intent
word_count
char_count
provider_used
model_used
impressions_90d
clicks_90d
pageviews_90d
sessions_90d
users_90d
engaged_sessions_90d
ai_sessions_90d
scroll_events_90d
days_with_impressions
days_with_sessions
impressions_last_30d
clicks_last_30d
sessions_last_30d
impressions_prev_30d
clicks_prev_30d
sessions_prev_30d
content_age_days
age_tier
age_tier_order
days_since_last_update
freshness_tier
word_count_tier
char_count_tier
ctr
avg_position
engagement_rate
scroll_rate
ai_traffic_pct
impression_tier
position_tier
trend_direction
trend_pct


In [50]:
df.head()

,content_id,client_id,search_volume,competition,competition_level,cpc,content_type,main_intent,word_count,char_count,...,char_count_tier,ctr,avg_position,engagement_rate,scroll_rate,ai_traffic_pct,impression_tier,position_tier,trend_direction,trend_pct
0,content_304f48230142,client_f369cb89fc,10.0,0.67,HIGH,2.05,keyword article,transactional,3221.0,20457.0,...,15000-25000,0.76,10.6,5.88,4.55,0.0,good,striking,down,-41.4
1,content_a1fb4e703a9e,client_4e07408562,90.0,0.01,LOW,0.05,keyword article,informational,2481.0,15562.0,...,15000-25000,0.05,20.3,0.00,10.00,0.0,good,page_3_5,down,-57.7
2,content_9aa793d4d895,client_7f2253d7e2,0.0,0.00,LOW,0.00,keyword article,informational,3515.0,23643.0,...,15000-25000,0.09,36.5,0.00,28.57,0.0,good,page_3_5,down,-60.9
3,content_331d6c4de07b,client_19581e27de,10.0,0.00,LOW,0.00,keyword article,commercial,NaN,NaN,...,NaN,0.49,6.2,1.28,3.45,0.0,good,page_1,stable,-13.8
4,content_d99b7a2d90ca,client_3fdba35f04,0.0,0.00,LOW,0.00,keyword article,informational,2803.0,17469.0,...,15000-25000,0.13,44.0,0.00,24.29,0.0,good,page_3_5,down,-34.7


## 1. My rule and its reason codes

*Write the rule in plain words first. Then the reason codes it can output.*

### Baseline rule

I will prioritize content for refresh when it shows a combination of staleness, declining trend, and search opportunity.

The rule uses:
- days_since_last_update as the staleness signal
- trend_direction as the performance-direction signal
- search_volume as the opportunity signal

A higher score means higher refresh priority. This is a decision-support baseline, not a claim that every high-scoring page requires a refresh.

In [51]:
# Signal check 1: staleness

df["staleness_bucket"] = pd.cut(
    df["days_since_last_update"],
    bins=[-1, 90, 180, 365, float("inf")],
    labels=["0-90", "91-180", "181-365", "365+"]
)

staleness_check = (
    df.groupby("staleness_bucket", observed=False)
      .agg(
          n=("content_id", "size"),
          avg_impressions_90d=("impressions_90d", "mean"),
          avg_ctr=("ctr", "mean"),
          avg_trend_pct=("trend_pct", "mean")
      )
      .reset_index()
)

display(staleness_check)

,staleness_bucket,n,avg_impressions_90d,avg_ctr,avg_trend_pct
0,0-90,20655,4219.161317,0.604856,0.707094
1,91-180,9171,7486.665140,0.238367,-15.683224
2,181-365,169,1206.893491,3.210828,-4.718462
3,365+,5,8.200000,20.000000,-96.166667


# Signal check 1: staleness

df["staleness_bucket"] = pd.cut(
    df["days_since_last_update"],
    bins=[-1, 90, 180, 365, float("inf")],
    labels=["0-90", "91-180", "181-365", "365+"]
)

staleness_check = (
    df.groupby("staleness_bucket", observed=False)
      .agg(
          n=("content_id", "size"),
          avg_impressions_90d=("impressions_90d", "mean"),
          avg_ctr=("ctr", "mean"),
          avg_trend_pct=("trend_pct", "mean")
      )
      .reset_index()
)

display(staleness_check)

In [52]:
# Signal check 2: trend direction

trend_check = (
    df.groupby("trend_direction", dropna=False)
      .agg(
          n=("content_id", "size"),
          avg_impressions_90d=("impressions_90d", "mean"),
          avg_ctr=("ctr", "mean"),
          avg_search_volume=("search_volume", "mean")
      )
      .reset_index()
)

display(trend_check)

,trend_direction,n,avg_impressions_90d,avg_ctr,avg_search_volume
0,down,16262,4919.097466,0.324138,133.376490
1,flat,1152,20.738715,1.376753,209.167483
2,new,2236,164.131038,1.296521,90.098039
3,stable,5962,9213.615230,0.517321,196.900070
4,up,4388,4716.081130,0.565533,211.049534


# Signal check 2: trend direction

trend_check = (
    df.groupby("trend_direction", dropna=False)
      .agg(
          n=("content_id", "size"),
          avg_impressions_90d=("impressions_90d", "mean"),
          avg_ctr=("ctr", "mean"),
          avg_search_volume=("search_volume", "mean")
      )
      .reset_index()
)

display(trend_check)

### Signal verdict — Trend direction

**CONFIRMED**

The observed bucket table shows that pages with a declining trend have weaker performance on the measured metrics than at least some non-declining groups. This supports using declining trend as a positive component of refresh priority.

## 2. Build the ranked queue (writes the CSV)

*Code the score, rank everything, write work/outputs/baseline_action_score.csv.*

In [53]:
import numpy as np

# Normalize the two numeric opportunity signals to a 0-100 scale.
staleness_score = (
    df["days_since_last_update"]
    / df["days_since_last_update"].max()
) * 100

search_score = (
    df["search_volume"].fillna(0)
    / df["search_volume"].max()
) * 100

# Give full trend points only to pages with a declining trend.
trend_score = np.where(
    df["trend_direction"].eq("down"),
    100,
    0
)

# One simple additive baseline score.
df["baseline_score"] = (
    0.40 * staleness_score
    + 0.30 * search_score
    + 0.30 * trend_score
)

# One reason code and one action label.
df["reason_code"] = "STALE_DECLINING_CONTENT"
df["action_label"] = "REFRESH_CONTENT"

# Rank highest score first.
queue = (
    df.sort_values(
        ["baseline_score", "content_id"],
        ascending=[False, True]
    )
    .reset_index(drop=True)
)

queue["rank"] = queue.index + 1

display(
    queue[
        [
            "rank",
            "content_id",
            "baseline_score",
            "reason_code",
            "action_label",
            "days_since_last_update",
            "search_volume",
            "trend_direction"
        ]
    ].head(20)
)

,rank,content_id,baseline_score,reason_code,action_label,days_since_last_update,search_volume,trend_direction
0,1,content_55a5b1c46474,70.000000,STALE_DECLINING_CONTENT,REFRESH_CONTENT,373,0.0,down
1,2,content_f6fdf87348f6,70.000000,STALE_DECLINING_CONTENT,REFRESH_CONTENT,373,0.0,down
2,3,content_1b4ec72dafd4,69.892761,STALE_DECLINING_CONTENT,REFRESH_CONTENT,372,0.0,down
3,4,content_f01216059a6a,65.924933,STALE_DECLINING_CONTENT,REFRESH_CONTENT,335,NaN,down
4,5,content_e2b702f4f92b,65.817694,STALE_DECLINING_CONTENT,REFRESH_CONTENT,334,NaN,down
5,6,content_454cc6654c6e,65.679842,STALE_DECLINING_CONTENT,REFRESH_CONTENT,104,60500.0,down
6,7,content_bf67a444faef,65.679842,STALE_DECLINING_CONTENT,REFRESH_CONTENT,104,60500.0,down
7,8,content_02b0d6e30129,63.610278,STALE_DECLINING_CONTENT,REFRESH_CONTENT,313,110.0,down
8,9,content_7a888d3d99c8,63.602170,STALE_DECLINING_CONTENT,REFRESH_CONTENT,313,90.0,down
9,10,content_94991fe6268c,63.569738,STALE_DECLINING_CONTENT,REFRESH_CONTENT,313,10.0,down


In [54]:
import os

output_dir = "/content/flyrank-ml-internship-haseeb/work/outputs"
os.makedirs(output_dir, exist_ok=True)

output_path = os.path.join(
    output_dir,
    "baseline_action_score.csv"
)

queue.to_csv(output_path, index=False)

print(f"Saved: {output_path}")
print(f"Rows written: {len(queue):,}")

Saved: /content/flyrank-ml-internship-haseeb/work/outputs/baseline_action_score.csv
Rows written: 30,000


In [55]:
import os
import pandas as pd

print("File exists:", os.path.exists(output_path))

check = pd.read_csv(output_path)

print("CSV shape:", check.shape)

display(
    check[
        [
            "rank",
            "content_id",
            "baseline_score",
            "reason_code",
            "action_label"
        ]
    ].head(10)
)

File exists: True
CSV shape: (30000, 49)


,rank,content_id,baseline_score,reason_code,action_label
0,1,content_55a5b1c46474,70.000000,STALE_DECLINING_CONTENT,REFRESH_CONTENT
1,2,content_f6fdf87348f6,70.000000,STALE_DECLINING_CONTENT,REFRESH_CONTENT
2,3,content_1b4ec72dafd4,69.892761,STALE_DECLINING_CONTENT,REFRESH_CONTENT
3,4,content_f01216059a6a,65.924933,STALE_DECLINING_CONTENT,REFRESH_CONTENT
4,5,content_e2b702f4f92b,65.817694,STALE_DECLINING_CONTENT,REFRESH_CONTENT
5,6,content_454cc6654c6e,65.679842,STALE_DECLINING_CONTENT,REFRESH_CONTENT
6,7,content_bf67a444faef,65.679842,STALE_DECLINING_CONTENT,REFRESH_CONTENT
7,8,content_02b0d6e30129,63.610278,STALE_DECLINING_CONTENT,REFRESH_CONTENT
8,9,content_7a888d3d99c8,63.602170,STALE_DECLINING_CONTENT,REFRESH_CONTENT
9,10,content_94991fe6268c,63.569738,STALE_DECLINING_CONTENT,REFRESH_CONTENT


In [56]:
import os
import pandas as pd

print("File exists:", os.path.exists(output_path))

check = pd.read_csv(output_path)

print("CSV shape:", check.shape)

display(
    check[
        [
            "rank",
            "content_id",
            "baseline_score",
            "reason_code",
            "action_label"
        ]
    ].head(10)
)

File exists: True
CSV shape: (30000, 49)


,rank,content_id,baseline_score,reason_code,action_label
0,1,content_55a5b1c46474,70.000000,STALE_DECLINING_CONTENT,REFRESH_CONTENT
1,2,content_f6fdf87348f6,70.000000,STALE_DECLINING_CONTENT,REFRESH_CONTENT
2,3,content_1b4ec72dafd4,69.892761,STALE_DECLINING_CONTENT,REFRESH_CONTENT
3,4,content_f01216059a6a,65.924933,STALE_DECLINING_CONTENT,REFRESH_CONTENT
4,5,content_e2b702f4f92b,65.817694,STALE_DECLINING_CONTENT,REFRESH_CONTENT
5,6,content_454cc6654c6e,65.679842,STALE_DECLINING_CONTENT,REFRESH_CONTENT
6,7,content_bf67a444faef,65.679842,STALE_DECLINING_CONTENT,REFRESH_CONTENT
7,8,content_02b0d6e30129,63.610278,STALE_DECLINING_CONTENT,REFRESH_CONTENT
8,9,content_7a888d3d99c8,63.602170,STALE_DECLINING_CONTENT,REFRESH_CONTENT
9,10,content_94991fe6268c,63.569738,STALE_DECLINING_CONTENT,REFRESH_CONTENT


## 3. Top-20 review

*For each of the top 20: action, reason code, confidence note, and what would make it wrong.*

### Top-20 review

The baseline assigns REFRESH_CONTENT to all top-20 rows because they meet the same rule and receive the STALE_DECLINING_CONTENT reason code.

The confidence notes below are intentionally cautious. A high score means the row matches the rule strongly; it does not prove that refreshing the page will improve performance.

**Rank 1 — content_55a5b1c46474**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 373 days since its last update and has a declining trend.
- What would make it wrong: Search volume is 0, so there is little observed search opportunity supporting a refresh.

**Rank 2 — content_f6fdf87348f6**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 373 days stale and has a declining trend.
- What would make it wrong: Search volume is 0, so the refresh opportunity may be limited.

**Rank 3 — content_1b4ec72dafd4**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 372 days stale and has a declining trend.
- What would make it wrong: The page may have limited search opportunity or the decline may have another cause.

**Rank 4 — content_f01216059a6a**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 335 days stale and trending down.
- What would make it wrong: Search volume is missing, so the opportunity signal is uncertain.

**Rank 5 — content_e2b702f4f92b**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 334 days stale and trending down.
- What would make it wrong: Search volume is missing, so the refresh opportunity is uncertain.

**Rank 6 — content_454cc6654c6e**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: High relative to the baseline. The page has 60,500 search volume and a declining trend.
- What would make it wrong: The traffic decline may be caused by seasonality, competition, or another external factor.

**Rank 7 — content_bf67a444faef**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: High relative to the baseline. The page has 60,500 search volume and a declining trend.
- What would make it wrong: The decline may not be caused by outdated content.

**Rank 8 — content_02b0d6e30129**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 313 days stale, has search volume of 110, and is trending down.
- What would make it wrong: The search opportunity is relatively small.

**Rank 9 — content_7a888d3d99c8**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 313 days stale and trending down.
- What would make it wrong: Search volume is only 90, so the expected opportunity may be limited.

**Rank 10 — content_94991fe6268c**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium-low. The page is 313 days stale and trending down.
- What would make it wrong: Search volume is only 10, making this a weak opportunity despite the staleness.

**Rank 11 — content_129753e3095f**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium-low. The page is 305 days stale and trending down.
- What would make it wrong: Search volume is only 10.

**Rank 12 — content_ab18b5811c02**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium-low. The page is 305 days stale and trending down.
- What would make it wrong: Search volume is only 10.

**Rank 13 — content_f488400fca67**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 305 days stale and trending down.
- What would make it wrong: Search volume is missing, so opportunity is uncertain.

**Rank 14 — content_07ce98c6085a**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 304 days stale and trending down.
- What would make it wrong: Search volume is missing, so the opportunity signal is uncertain.

**Rank 15 — content_7736e6144f3b**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 304 days stale and trending down.
- What would make it wrong: Search volume is missing, so the opportunity signal is uncertain.

**Rank 16 — content_84d12054c0c0**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 304 days stale and trending down.
- What would make it wrong: Search volume is missing, so the opportunity signal is uncertain.

**Rank 17 — content_df1fa766cac2**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 304 days stale and trending down.
- What would make it wrong: Search volume is missing, so the opportunity signal is uncertain.

**Rank 18 — content_72496874f806**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 301 days stale and trending down.
- What would make it wrong: Search volume is missing from the displayed ranking, so the opportunity evidence is incomplete.

**Rank 19 — content_7f116ae1f6f5**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 301 days stale and trending down.
- What would make it wrong: Search volume is missing from the displayed ranking, so the opportunity evidence is incomplete.

**Rank 20 — content_8f2c815af658**
- Action: REFRESH_CONTENT
- Reason code: STALE_DECLINING_CONTENT
- Confidence: Medium. The page is 301 days stale and trending down.
- What would make it wrong: Search volume is missing from the displayed ranking, so the opportunity evidence is incomplete.

## 4. Weak picks + leakage check

*Which picks look wrong and why? Confirm no product flags or future windows leaked in.*

In [57]:
weak_picks = queue.tail(10).copy()

display(
    weak_picks[
        [
            "rank",
            "content_id",
            "baseline_score",
            "action_label",
            "reason_code",
            "days_since_last_update",
            "search_volume",
            "trend_direction"
        ]
    ]
)

,rank,content_id,baseline_score,action_label,reason_code,days_since_last_update,search_volume,trend_direction
29990,29991,content_b5fb35404aed,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,flat
29991,29992,content_bb600f317035,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,new
29992,29993,content_bfea9068e6f2,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,flat
29993,29994,content_c22952dee240,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,flat
29994,29995,content_c53b8d4bce75,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,new
29995,29996,content_d269ebb68607,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,flat
29996,29997,content_dc37f9bc33fb,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,stable
29997,29998,content_e13946b1656b,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,flat
29998,29999,content_ebcbcaf7b641,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,up
29999,30000,content_ef93c98775d5,0.107239,REFRESH_CONTENT,STALE_DECLINING_CONTENT,1,0.0,new


### Weak picks

The weakest-ranked pages are lower priority under this baseline because they have a weaker combination of staleness, search opportunity, and declining trend.

These are not necessarily pages that should never be refreshed. A low score only means that the simple baseline does not prioritize them.

The main weakness observed in the top-ranked results is that some pages receive high scores because they are very stale and trending down even when search volume is zero or missing. This means the baseline can over-prioritize stale pages with limited or uncertain search opportunity.

In [58]:
baseline_inputs = [
    "days_since_last_update",
    "search_volume",
    "trend_direction"
]

print("Inputs used by baseline:")
for col in baseline_inputs:
    print(f"- {col}")

print("\nLeakage checks:")
print("- Product/FlyRank flags used: NO")
print("- Future-window labels used: NO")
print("- Target/label-derived inputs used: NO")
print("- Client ID used for scoring: NO")

Inputs used by baseline:
- days_since_last_update
- search_volume
- trend_direction

Leakage checks:
- Product/FlyRank flags used: NO
- Future-window labels used: NO
- Target/label-derived inputs used: NO
- Client ID used for scoring: NO


### Leakage check

No FlyRank product flags were used as scoring inputs.

No future-window outcome or target/label was used.

The score uses only:
- days_since_last_update
- search_volume
- trend_direction

Client identifiers are not used in the score.

The baseline is intended as decision-support and should be interpreted as an observed, directional ranking rather than a causal or predictive claim.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.